pkg/operator/cockroachdb/controller.go

```diff
@@ -0,0 +1,573 @@
+										Port: intstr.FromString(httpPortName),
+									},
+								},
+								InitialDelaySeconds: int32(10),
+								PeriodSeconds:       int32(5),
+								FailureThreshold:    int32(2),
+							},
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      volumeNameDataDir,
+									MountPath: "/cockroach/cockroach-data",
+								},
+							},
+							Env: []v1.EnvVar{
+								{
+									Name:  envVarChannel,
+									Value: envVarChannelVal,
+								},
+							},
+							Command: []string{
+								"/bin/bash",
+								"-ecx",
+								createCommand(cluster),
+							},
+						},
+					},
+					//  No pre-stop hook is required, a SIGTERM plus some time is all that's needed for graceful shutdown of a node.
+					TerminationGracePeriodSeconds: &terminationGracePeriodSeconds,
+					Volumes: []v1.Volume{
+						{
+							Name: volumeNameDataDir,
+							VolumeSource: v1.VolumeSource{
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: volumeNameDataDir,
+								},
+							},
+						},
+					},
+				},
+			},
+			PodManagementPolicy: appsv1beta1.ParallelPodManagement,
+			UpdateStrategy: appsv1beta1.StatefulSetUpdateStrategy{
+				Type: appsv1beta1.RollingUpdateStatefulSetStrategyType,
+			},
+			VolumeClaimTemplates: []v1.PersistentVolumeClaim{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      volumeNameDataDir,
+						Namespace: cluster.namespace,
+					},
+					Spec: v1.PersistentVolumeClaimSpec{
+						AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+						Resources: v1.ResourceRequirements{
+							Requests: v1.ResourceList{
+								v1.ResourceStorage: cluster.spec.VolumeSize,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if _, err := c.context.Clientset.AppsV1beta1().StatefulSets(cluster.namespace).Create(statefulSet); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return err
+		}
+		logger.Infof("stateful set %s already exists in namespace %s", statefulSet.Name, statefulSet.Namespace)
+	} else {
+		logger.Infof("stateful set %s created in namespace %s", statefulSet.Name, statefulSet.Namespace)
+	}
+
+	return nil
+}
+
+func (c *ClusterController) initCluster(cluster *cluster) error {
+	hostFlag := fmt.Sprintf("--host=%s", createQualifiedReplicaServiceName(0, cluster.namespace))
+	out, err := c.context.Executor.ExecuteCommandWithCombinedOutput(false, "cockroachdb init",
+		"/cockroach/cockroach", "init", "--insecure", hostFlag)
+
+	var outcome string
+	if err != nil {
+		outcome = "failed"
+	} else {
+		outcome = "succeeded"
+	}
+
+	logger.Infof("cluster init %s for namespace %s: %s", outcome, cluster.namespace, out)
+	return err
+}
+
+func validateClusterSpec(spec cockroachdbv1alpha1.ClusterSpec) error {
+	if spec.Storage.NodeCount < 1 {
+		return fmt.Errorf("invalid node count: %d. Must be at least 1", spec.Storage.NodeCount)
+	}
+
+	if err := validatePercentValue(spec.CachePercent, "cache"); err != nil {
+		return err
```

Comment:
Perhaps log why its invalid rather than just returning an error. Would make life easier when debugging.

Reply:
The `validatePercentValue` function includes the value and that it must be between 0 and 100 into the message for the error object, which then gets logged back in the `onAdd` function.  The reason for being invalid should make it to the logs IMO, help me understand if I'm missing something more.